asst2/CODE/memory_finder.c

```diff
@@ -0,0 +1,438 @@
+		//update HT, PT, and metadata is page is allocated to process
+		if ( !pg_block_occupied(pdata) || pdata->pid == curr_id ) {
+			//add the actual index and where process thinks the page is to HT
+			ht_put(ht_space, curr_id, (ht_val)process_index, actual_index);
+			//if last page to alloc, set metadata for the free segment following the overflow segment
+			if (pages_alloc == 1 && ovf_len != num_segments && swap_array_index != 0){
+				int curr_seg = dm_block_size(start);
+				dm_write_metadata(start+ovf_len*SEGMENTSIZE,curr_seg-ovf_len, NOT_OCC, dm_is_last_segment(start));
+			}
+			//if first page in allocation set page data and set segment data of the last
+			if (swap_array_index == 0) {
+//				printf("page mallc\n");
+				int is_ovf = (pages_alloc == 1) ? NOT_OVF : OVF;
+				pg_write_pagedata(pdata, curr_id, OCC, is_ovf, process_index, pages_alloc);
+				dm_allocate_block(start,ovf_len);
+			}
+			else { 	//if not 1st page, write pagedata for process owning page and where process thinks it is,
+				pg_write_pagedata(pdata, curr_id, OCC, NOT_OVF, process_index, 1);
+			}
+			//add the free page allocated into array for swapping to correct location later
+			pages_actual_index[swap_array_index] = actual_index;
+			swap_array_index++;
+			pages_alloc--;
+			process_index++;
+		}
+
+		actual_index++;
+	}
+	//swap newly allocated pages to where the process thinks they are
+	//and set return pointer to be returned from first page newly allocated
+	void* ret;
+	for (int i = 0; i < swap_len; i++){
+		pagedata* p = (pagedata*)myblock+pages_actual_index[i];
+		if (i == 0) {
+			ret = mem_space+pg_index(p)*page_size+ret_seg_index*SEGMENTSIZE+1;
+		}
+		swap(pages_actual_index[i], pg_index(p) );
+	}
+
+	return ret;
+}
+
+/**
+ * Services requests that are within a page or outsources to page_allocate() multiple page allocations
+ * @param size			size of malloc request in bytes
+ * @param curr_id		process making request
+ */
+void* segment_allocate(size_t size, my_pthread_t curr_id){
+	int first_page_index;
+	int last_page_index;
+	int free_page;
+	int num_free;
+	find_contig_space(curr_id, &free_page, &num_free, &first_page_index, &last_page_index);
+	size_t segments_alloc = size / SEGMENTSIZE + 1;
+	int pages_alloc = (segments_alloc/num_segments+1);
+
+	int alloc_page_index; //page index according the process where an allocation could start
+	int alloc_seg_index = -1; //segment index within alloc_page_index page where allocation could start
+	int contig_free_space = 0; //from start index, to current index how much contiguous free space available
+	int process_index = first_page_index;
+	uint8_t ovf_len = 0;
+	//traverse memory to find the most current current contigious free block that can fit the size requested
+	while (process_index <= last_page_index) {
+		int actual_index = ht_get(ht_space, curr_id, (ht_key)process_index);
+		pagedata* pdata = (pagedata*)myblock + actual_index;
```

Comment:
@Victor-C-Zhang Check it out: here I use the` find_contig_space()` (line 178) to find the start and end of process' view of memory space, then I use that to traverse the memory space, but rather than reading from where the process thinks its page is, i simply use that value to find where the page actually is and read from there. Once I read from there, I increment the index to the next index according to process view, and again use this index to find where the next page actually is. I hope that makes it a bit clearer.

Reply:
yes, this makes sense now thanks. however, I'm not sure this will work if we can swap to disk. maybe we'll need to put it in the TCB?

Reply:
hmm put into TCB? put what in the tcb ?

Reply:
first/last page numbers. since we can't be sure they'll be swapped into main memory

Reply:
yea that we can do, we would need one for library too